### ISSUE TITLE
Doesn't seem to play well with Jquery 3

### ISSUE THREAD

**Issue by ansonlouis** [tbranyen/backbone.layoutmanager#507]
We're trying to upgrade to jQuery 3 at my work and it seems layoutmanager doesn't play well with it. 

I've found the issue comes down to jQuery adding Promise/A+ compliance, which changes the behavior of `.then()` (forced async, only one resolve parameter, etc).

This change also causes `$.when()` to assume any passed object with a `.then()` method to be a "promise". I believe this is where the issue with layoutmanager and jQuery3 lies, since `.then()` is proxied onto the layoutmanager view (line :265) but is not a promise. The end cause is that subviews tend to not be rendered.

jQuery stated in their [migration guide](https://jquery.com/upgrade-guide/3.0/) that old `$.Deferred` methods such as `done` and `fail` maintain parity with previous jQuery versions and are not subject to the breaking changes. So, changing instances of `then` to `done` throughout layoutmanager seems to be a reasonable solution to me, seeing as how I don't see any instances in which a call to `then` returns a filtered value. I also dont see deferreds being returned anywhere, meaning (for all practical purposes) `done` would work just fine.  At the very least, the proxied `then` method should change to `done` and any place it was being called.

**Comment by atomictag** [comment]
Yes that's correct. You should provide a custom version of `when` passing it in as a configuration option, something like:
```
  when: function(promises) {
    var deferred = this.deferred();
    var resolve  = _.after(promises.length, deferred.resolve);
    _.each(promises, function(promise) {
      if(promise && promise.then) promise.done(resolve).fail(deferred.reject);
      else resolve();
    });
    return deferred.promise();
    // **** this was the original
    // return $.when.apply(null, promises);
  },
```
Note that Promise/A+ compliance in jQ3 means that `then` is now *always* asynchronous (emulated with `setTimeout` with a 0 interval when needed). This is unfortunate in some cases. If thar causes you troubles you need to replace all occurrences of `.then` in the library (such as `promise().then()` with `.done()`. `done` is not a Promise/A+ artefact, so it does not follow the behaviour of `then`

**Comment by atomictag** [comment]
there are a few places where `.when` is used in the code. You don't actually _change_ `when` - it's a very clean thing you can do without changing the library code, because you can pass it in LM's configuration.

**Comment by ansonlouis** [comment]
I'm not sure I follow. I realize what you're suggesting might fix the problem. I'm submitting an issue to the library because it _does not work with jQuery 3_. The ultimate solution to this problem isn't to pass in my own version of `when` to the view while using the library. The solution is to _fix the library_ so that it works with jquery 3. In doing so, nobody has to do what you suggest and it works as it is suppose to work.

What you're suggesting is analogous to telling people to _clear their cookies_ when your app doesn't work, instead of fixing the problem so that clearing the cookies _is not required_ by the user.

**Comment by ansonlouis** [comment]
Yeah it's no problem, I do appreciate it. My goal with opening up the issue on this repo is to fix the library so no one else needs to patch it.
